consent/handler.go

```diff
@@ -1049,6 +1053,133 @@ func (h *Handler) getOAuth2LogoutRequest(w http.ResponseWriter, r *http.Request,
+//
+// Accepts a device grant user_code request
+//
+//	Consumes:
+//	- application/json
+//
+//	Produces:
+//	- application/json
+//
+//	Schemes: http, https
+//
+//	Responses:
+//	  200: oAuth2RedirectTo
+//	  default: errorOAuth2
+func (h *Handler) acceptUserCodeRequest(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	ctx := r.Context()
+
+	challenge := r.URL.Query().Get("device_challenge")
+	if challenge == "" {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHint(`Query parameter 'device_challenge' is not defined but should have been.`)))
+		return
+	}
+
+	var reqBody flow.AcceptDeviceUserCodeRequest
+	d := json.NewDecoder(r.Body)
+	d.DisallowUnknownFields()
+	if err := d.Decode(&reqBody); err != nil {
+		h.r.Writer().WriteErrorCode(w, r, http.StatusBadRequest, errorsx.WithStack(err))
+		return
+	}
+
+	if reqBody.UserCode == "" {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHint("Field 'user_code' must not be empty.")))
+		return
+	}
+
+	cr, err := h.r.ConsentManager().GetDeviceUserAuthRequest(ctx, challenge)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(err))
+		return
+	}
+
+	f, err := h.decodeFlowWithClient(ctx, challenge, flowctx.AsDeviceChallenge)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	userCodeSignature, err := h.r.RFC8628HMACStrategy().UserCodeSignature(r.Context(), reqBody.UserCode)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithHint(`'user_code' signature could not be computed`)))
+		return
+	}
+	userCodeRequest, err := h.r.OAuth2Storage().GetUserCodeSession(r.Context(), userCodeSignature, nil)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrNotFound.WithWrap(err).WithHint(`'user_code' session not found`)))
+		return
+	}
+	err = h.r.RFC8628HMACStrategy().ValidateUserCode(ctx, userCodeRequest, reqBody.UserCode)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrTokenExpired.WithWrap(err).WithHint(`'user_code' has expired`)))
+		return
+	}
+
+	p := flow.HandledDeviceUserAuthRequest{
+		ID:                  f.DeviceChallengeID.String(),
+		RequestedAt:         cr.RequestedAt,
+		HandledAt:           sqlxx.NullTime(time.Now().UTC()),
+		Client:              userCodeRequest.GetClient().(*client.Client),
+		DeviceCodeRequestID: userCodeRequest.GetID(),
+		RequestedScope:      []string(userCodeRequest.GetRequestedScopes()),
+		RequestedAudience:   []string(userCodeRequest.GetRequestedAudience()),
+	}
+
+	// Append the client_id to the original RequestURL, as it is needed for the login flow
+	reqURL, err := url.Parse(f.RequestURL)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(err))
+		return
+	}
+	if reqURL.Query().Get("client_id") == "" {
+		q := reqURL.Query()
+		q.Add("client_id", userCodeRequest.GetClient().GetID())
+		reqURL.RawQuery = q.Encode()
+	}
```

Comment:
Should we not **always** use the client from the usercode request? Why is it possible to override the client here? This is a bit confusing to me as to why we need this.

Reply:
The initial request should not contain the `client_id`. We don't know the client_id when the user gets to the device verification endpoint, we need to fetch the user_code from the database to get it.

The only reason we are adding the `client_id` to the request is to make the existing endpoints work without too many changes. The `client_id` is parsed in https://github.com/nsklikas/hydra/blob/canonical-master/consent/strategy_default.go#L1211 to construct an authorization request that will be used to do the rest of the flow. If we didn't have this hack, we would have to refactor part of the existing default_strategy. IMHO this is work that should be done after this PR is merged.

I will add a comment to make this clearer.

Reply:
Ok I see, is it possible that we end up here where the user code / authorization flow and the client id could mismatch?